src/extensions/dnatco/confal-pyramids/representation.ts

```diff
@@ -0,0 +1,203 @@
+    Vec3.add(mp, mp, w);
+}
+
+function shiftVertex(vec: Vec3, ref: Vec3, scale: number) {
+    Vec3.sub(w, vec, ref);
+    Vec3.scale(w, w, scale);
+    Vec3.add(vec, vec, w);
+}
+
+const ConfalPyramidsMeshParams = {
+    ...UnitsMeshParams
+};
+type ConfalPyramidsMeshParams = typeof ConfalPyramidsMeshParams;
+
+function createConfalPyramidsIterator(structureGroup: StructureGroup): LocationIterator {
+    const { structure } = structureGroup;
+    const modelNum = structure.model.modelNum;
+    const unit = structureGroup.group.units[0];
+    const location = StructureElement.Location.create(structure, unit); // Dummy location
+
+    // Always returns dummy location
+    const noopIterator = LocationIterator(0, 1, (groupIndex: number, instanceIndex: number) => {
+        return location;
+    });
+
+    if (!Unit.isAtomic(unit)) return noopIterator;
+
+    const mapping = ConfalPyramidsUtil.getPyramidsMapping(structure.model, modelNum);
+
+    if (!mapping.groupMap.has(unit.invariantId)) return noopIterator;
+    const unitMap = mapping.groupMap.get(unit.invariantId)!;
+
+    const getLocation = (groupIndex: number, instanceIndex: number) => {
+        if (groupIndex >= unitMap.length)
+            throw new Error('Invalid groupIndex');
+        return unitMap[groupIndex];
+    };
+    return LocationIterator(unitMap.length, 1, getLocation);
+}
+
+function createConfalPyramidsMesh(ctx: VisualContext, unit: Unit, structure: Structure, theme: Theme, props: PD.Values<ConfalPyramidsMeshParams>, mesh?: Mesh) {
+    if (!Unit.isAtomic(unit)) return Mesh.createEmpty(mesh);
+
+    const mapping = ConfalPyramidsUtil.getPyramidsMapping(unit.model, unit.model.modelNum);
+    if (mapping.pyramids.length === 0) return Mesh.createEmpty(mesh);
+
+    const mb = MeshBuilder.createState(512, 512, mesh);
+    mb.currentGroup = 0;
+
+    const unitId = unit.invariantId;
+    if (!mapping.groupMap.has(unitId)) mapping.groupMap.set(unitId, new Array<StructureElement.Location>());
+    const unitMap = mapping.groupMap.get(unitId)!;
+    unitMap.splice(0, unitMap.length);
+
+    const handler = (pyramid: Pyramid, index: number, first: ConfalPyramidsUtil.FirstResidueAtoms, second: ConfalPyramidsUtil.SecondResidueAtoms, locFirst: StructureElement.Location, locSecond: StructureElement.Location) => {
+        if (index === -1)
+            throw new Error('Invalid pyramid index');
+
+        const scale = (pyramid.confal_score - 20.0) / 100.0;
+        const O3 = first.O3;
+        const OP1 = second.OP1, OP2 = second.OP2, O5 = second.O5, P = second.P;
+
+        shiftVertex(O3, P, scale);
+        shiftVertex(OP1, P, scale);
+        shiftVertex(OP2, P, scale);
+        shiftVertex(O5, P, scale);
+        calcMidpoint(mp, O3, O5);
+
+        let pb = PrimitiveBuilder(3);
+        pb.add(O3, OP1, OP2);
+        pb.add(O3, mp, OP1);
+        pb.add(O3, OP2, mp);
+        MeshBuilder.addPrimitive(mb, t, pb.getPrimitive());
+
+        unitMap.push(locFirst);
```

Comment:
How about creating the `unitMap` or something similar in `createPyramidsFromCif`? It does not seem right to me to have it uninitialized until a mesh is created. I would probably create an array like `confalResidue: ArraLike<ConfalIndex | -1>` with entries for each ResidueIndex using -1 to indicate nothing is there for a given residue. ConfalIndex would point to another array that contains all data needed for a residues pyramid.

Reply:
The code needs some way how to assign `groupId` to `Location` in order to figure out colors of the pyramids efficiently. Since I will be making some adjustments to the code anyway I will see if I can accommodate your suggestion in some way.